web/sharings/move.go

```diff
@@ -0,0 +1,468 @@
+	if err != nil {
+		return files.WrapVfsError(err)
+	}
+	u, err := url.Parse(sourceInstanceURL)
+	if err != nil {
+		return files.WrapVfsError(err)
+	}
+
+	bearer, err := extractBearerToken(s)
+	if err != nil {
+		return err
+	}
+	srcClient := NewRemoteClient(u, bearer)
+
+	srcFile, err := srcClient.GetFileByID(fileID)
+	if err != nil {
+		return files.WrapVfsError(err)
+	}
+
+	newFileDoc, err := createFileDocFromRemoteFile(srcFile, destDir.DocID)
+	if err != nil {
+		return files.WrapVfsError(err)
+	}
+
+	fd, err := inst.VFS().CreateFile(newFileDoc, nil)
+	if err != nil {
+		return files.WrapVfsError(err)
+	}
+
+	rc, err := srcClient.DownloadByID(fileID)
+	if err != nil {
+		// Best-effort close to avoid leaking descriptors on error paths
+		_ = fd.Close()
+		return files.WrapVfsError(err)
+	}
+	defer rc.Close()
+
+	// Use a small buffer to reduce allocations during large copies
+	if _, err := io.CopyBuffer(fd, rc, make([]byte, copyBufferSize)); err != nil {
+		// Best-effort close to avoid leaking descriptors on error paths
+		_ = fd.Close()
+		return files.WrapVfsError(err)
+	}
+	err = fd.Close()
+	if err != nil {
+		return err
+	}
+	if err := srcClient.PermanentDeleteByID(fileID); err != nil {
+		inst.Logger().WithNamespace("move").Warnf("Could not delete source file: %v", err)
+	}
+	obj := files.NewFile(newFileDoc, inst, nil)
+	return jsonapi.Data(c, http.StatusCreated, obj, nil)
+}
+
+// MoveHandler is a unified endpoint to move a file between different locations.
+// Accepted inputs (JSON body preferred, query params supported for compatibility):
+// - Common: file-id, dir-id
+// - Source: source-instance + source-sharing_id (both together for shared drives)
+// - Destination: dest-instance + dest-sharing_id (both together for shared drives)
+// For personal drives, only sharing_id is provided (instance is empty).
+// At least one sharing_id must be provided.
+//
+// Also supports a nested JSON structure with source and target objects:
+//
+//		{
+//		  "source": {
+//		    "instance": "https://alice.localhost:8080",
+//		    "file_id": "file123",
+//		    "sharing_id": "sharing123",
+//		    "dir_id": "dir456"
+//		  },
+//		  "dest": {
+//		    "instance": "https://alice.localhost:8080",
+//	        "sharing_id": "sharing123",
+//		    "dir_id": "dir456"
+//		  }
+//		}
+func MoveHandler(c echo.Context) error {
+	inst := middlewares.GetInstance(c)
+
+	var req moveRequest
+	if err := c.Bind(&req); err != nil {
+		return jsonapi.BadRequest(errors.New("error unmarshalling request"))
+	}
+	if err := validateMoveRequest(req); err != nil {
+		return err
+	}
+
+	// Authorization: if at least one side is not a shared drive, require whole-type permission
+	if req.Source.SharingID == "" || req.Dest.SharingID == "" {
+		if err := middlewares.AllowWholeType(c, permission.PATCH, consts.Files); err != nil {
+			return err
+		}
+	}
+
+	sourceInstance, err := resolveInstanceOrCurrent(req.Source.Instance, inst)
```

Comment:
If the source instance is not on the same stack as the local instance, it will return an error here